From aed082e112fa2e1502fc3e0e5876723a0da7c67b Mon Sep 17 00:00:00 2001
From: svenkatachalam <svenkatachalam@maxlinear.com>
Date: Thu, 22 Aug 2024 13:11:56 +0800
Subject: [PATCH] WLANRTSYS-82831 UCI config to disable beacon protection.

FIX:
Introduced UCI config to disable beacon protection,
as beacon protection is enabled by default for 11BE mode.

FIXUP to WLANRTSYS-60843 MLO changes: AP MLD configuration
---
 hostapd/mxl_config.c | 20 ++++++++++++++++++--
 src/ap/mxl_config.h  |  3 +++
 2 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/hostapd/mxl_config.c b/hostapd/mxl_config.c
index 5498c6c9f..887c4285c 100644
--- a/hostapd/mxl_config.c
+++ b/hostapd/mxl_config.c
@@ -1918,6 +1918,15 @@ int mxl_hostapd_config_fill (struct hostapd_config *conf,
 		mxl_conf->zwdfs_debug_chan = val;
 	} else if (os_strcmp(buf, "veriwave_test") == 0) {
 		mxl_bss_conf->veriwave_test = atoi(pos);
+#ifdef CONFIG_IEEE80211BE
+	} else if (os_strcmp(buf, "disable_beacon_prot") == 0) {
+		int val = atoi(pos);
+		if (!IS_VALID_BOOL_RANGE(val)) {
+			wpa_printf(MSG_ERROR, "Line %d: invalid disable_beacon_prot value %d", line, val);
+			return 0;
+		}
+		mxl_bss_conf->disable_beacon_prot = val;
+#endif /* CONFIG_IEEE80211BE */
 	} else
 		return 0;
 
@@ -2073,11 +2082,18 @@ int mxl_hostapd_config_check(struct hostapd_config *conf, int full_config)
 int mxl_hostapd_config_check_bss(struct hostapd_bss_config *bss, struct hostapd_config *conf, int full_config)
 {
 #ifdef CONFIG_IEEE80211BE
-	if (full_config && conf->ieee80211be && !bss->disable_11be &&
-		((bss->wpa_key_mgmt & WPA_KEY_MGMT_OWE) || wpa_key_mgmt_sae(bss->wpa_key_mgmt))) {
+	if (full_config && conf->ieee80211be && !bss->disable_11be) {
+		if (bss->mxl_bss_conf.disable_beacon_prot) {
+			bss->beacon_prot = 0;
+			wpa_printf(MSG_WARNING,
+				"Force disabling beacon protection for the BSS: %s", bss->iface);
+		}
+
+		if ((bss->wpa_key_mgmt & WPA_KEY_MGMT_OWE) || wpa_key_mgmt_sae(bss->wpa_key_mgmt)) {
 			bss->rsn_pairwise |= WPA_CIPHER_GCMP_256;
 			wpa_printf(MSG_WARNING,
 				"GCMP-256 cipher suite is force-enabled due to 802.11be is enabled");
+		}
 	}
 #endif /* CONFIG_IEEE80211BE */
 	return 0;
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index dca439229..e419bd463 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -405,6 +405,9 @@ struct mxl_hostapd_bss_config {
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 	int rnr_auto_update;
 	bool veriwave_test;
+#ifdef CONFIG_IEEE80211BE
+	bool disable_beacon_prot;
+#endif /* CONFIG_IEEE80211BE */
 };
 
 struct hostapd_config;
-- 
2.43.0

